src/storage-client/src/source/persist_source.rs

```diff
@@ -295,3 +266,54 @@ impl PendingWork {
         true
     }
 }
+
+pub fn mfp<G>(
+    collection: Collection<G, Result<Row, DataflowError>, Diff>,
+    name: &str,
+    until: Antichain<Timestamp>,
+    mfp: MfpPlan,
+) -> Collection<G, Result<Row, DataflowError>, Diff>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    // Re-used state for processing and building rows.
+    let mut datum_vec = mz_repr::DatumVec::new();
+    let mut row_builder = Row::default();
+
+    collection
+        .inner
+        .unary(Pipeline, &format!("persist_source::mfp({name})"), |_, _| {
+            let mut vector = Vec::new();
+            move |input, output| {
+                input.for_each(|time, data| {
+                    data.swap(&mut vector);
+                    let mut session = output.session(&time);
+                    for (row, ts, diff) in vector.drain(..) {
+                        match row {
+                            Ok(row) => {
+                                let arena = mz_repr::RowArena::new();
+                                let mut datums_local = datum_vec.borrow_with(&row);
+                                for result in mfp.evaluate(
+                                    &mut datums_local,
+                                    &arena,
+                                    ts,
+                                    diff,
+                                    |time| !until.less_equal(time),
+                                    &mut row_builder,
+                                ) {
+                                    session.give(match result {
+                                        Ok((r, t, d)) => (Ok(r), t, d),
+                                        Err((e, t, d)) => (Err(e), t, d),
```

Comment:
Here we'd have a great opportunity to not multiplex the ok and error streams. In most subsequent operators they're demuxed again, so it'd be better to keep them separate in the first place. However, it's fine to leave it as follow-up work.